subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -173,27 +192,40 @@ private String getFilename(Path dir) {
 
         @Override
         public FileVisitResult visitFile(Path file, @Nullable BasicFileAttributes attrs) {
-            String name = stringInterner.intern(file.getFileName().toString());
+            String name = intern(file.getFileName().toString());
             if (isAllowed(file, name, false, attrs, builder.getRelativePath())) {
                 if (attrs == null) {
                     throw new FileSnapshottingException(String.format("Cannot read file '%s': not authorized.", file));
                 }
-                if (attrs.isSymbolicLink()) {
-                    // when FileVisitOption.FOLLOW_LINKS, we only get here when link couldn't be followed
-                    throw new FileSnapshottingException(String.format("Could not list contents of '%s'. Couldn't follow symbolic link.", file));
-                }
-                addFileSnapshot(file, name, attrs);
+                FileSystemLocationSnapshot snapshot = getSnapshot(file, name, attrs);
+                builder.visit(snapshot);
             }
             return FileVisitResult.CONTINUE;
         }
 
+        private FileSystemLocationSnapshot getSnapshot(Path file, String name, BasicFileAttributes attrs) {
+            String absolutePath = intern(file.toString());
+            try {
+                HashCode hash = hasher.hash(file.toFile(), attrs.size(), attrs.lastModifiedTime().toMillis());
+                FileMetadata fileMetadata = FileMetadata.from(attrs);
+                return new RegularFileSnapshot(absolutePath, name, hash, fileMetadata);
+            } catch (UncheckedIOException e) {
+                LOGGER.debug("Could not read file path '{}'.", file, e);
+                return new MissingFileSnapshot(absolutePath, name, FileType.Unavailable);
+            }
+        }
+
         @Override
         public FileVisitResult visitFileFailed(Path file, IOException exc) {
             // File loop exceptions are ignored. When we encounter a loop (via symbolic links), we continue
             // so we include all the other files apart from the loop.
             // This way, we include each file only once.
-            if (isNotFileSystemLoopException(exc) && isAllowed(file, file.getFileName().toString(), false, null, builder.getRelativePath())) {
-                throw new UncheckedIOException(String.format("Could not read path '%s'.", file), exc);
+            if (isNotFileSystemLoopException(exc)) {
+                String name = file.getFileName().toString();
```

Comment:
Should we intern the file name here as well?

Reply:
Thought about it as well, but since I'm not sure why we're interning here, I don't know.
I'll intern, since you do seem to understand :)